nucypher/blockchain/eth/signers/hardware.py

```diff
@@ -0,0 +1,243 @@
+        self.__addresses = dict()  # track derived addresses
+        self.__load_addresses()
+
+    #
+    # Internal
+    #
+
+    def __get_address_path(self, index: int = None, checksum_address: str = None) -> List[H_]:
+        """Resolves a checksum address into an HD path and returns it."""
+        if index is not None and checksum_address:
+            raise ValueError("Expected index or checksum address; Got both.")
+        elif index is not None:
+            hd_path = parse_path(f"{self.DERIVATION_ROOT}/{index}")
+        else:
+            try:
+                hd_path = self.__addresses[checksum_address]
+            except KeyError:
+                raise RuntimeError(f"{checksum_address} was not loaded into the device address cache.")
+        return hd_path
+
+    @handle_trezor_call
+    def __get_address(self, index: int = None, hd_path: Address = None, show_display: bool = True) -> str:
+        """Resolves a trezorlib HD path into a checksum address and returns it."""
+        if not hd_path:
+            if index is None:
+                raise ValueError("No index or HD path supplied.")  # TODO: better error handling here
+            hd_path = self.__get_address_path(index=index)
+        address = ethereum.__get_address(client=self.__client, n=hd_path, show_display=show_display)
+        return address
+
+    def __load_addresses(self) -> None:
+        """
+        Derive trezor addresses up to ADDRESS_CACHE_SIZE relative to
+        the calculated base path and internally cache them.
+        """
+        for index in range(self.ADDRESS_CACHE_SIZE):
+            hd_path = self.__get_address_path(index=index)
+            address = self.__get_address(hd_path=hd_path, show_display=False)
+            self.__addresses[address] = hd_path
+
+    @staticmethod
+    def _format_transaction(transaction_dict: dict) -> dict:
+        """
+        Handle Web3.py -> Trezor native transaction field formatting
+        # https://web3py.readthedocs.io/en/latest/web3.eth.html#web3.eth.Eth.sendRawTransaction
+        """
+        assert_valid_fields(transaction_dict)
+        trezor_transaction_keys = {'gas': 'gas_limit', 'gasPrice': 'gas_price', 'chainId': 'chain_id'}
+        trezor_transaction = dict(apply_key_map(trezor_transaction_keys, transaction_dict))
+        return trezor_transaction
+
+    @handle_trezor_call
+    def __sign_transaction(self, n: List[int], trezor_transaction: dict) -> Tuple[bytes, bytes, bytes]:
+        """Internal wrapper for trezorlib transaction signing calls"""
+        v, r, s = trezorlib.ethereum.sign_tx(client=self.__client, n=n, **trezor_transaction)
+        return v, r, s
+
+    #
+    # Trezor Signer API
+    #
+
+    @classmethod
+    def from_signer_uri(cls, uri: str) -> 'TrezorSigner':
+        """Return a trezor signer from URI string i.e. trezor:///my/trezor/path """
+        if uri != cls.URI_SCHEME:
+            raise cls.InvalidSignerURI(f'{uri} is not a valid trezor URI scheme')
```

Comment:
Also, we can be more helpful here indicating what's the expected trezor URI scheme

Reply:
I agree - we need to decide on what a Trezor URI even *is*.

Reply:
(For some reason the related discussion we had on L152 is lost, continuing here). 
See what geth does as the output of `personal_listWallets`:
```
> personal.listWallets
INFO [09-18|19:48:22.776] USB wallet discovered new account        url=trezor://1209:53c1:01 address=0xfoobar path=m/44'/60'/0'/0/0 balance=0 nonce=0
[{
    accounts: [{
        address: "0xfoobar",
        url: "trezor://1209:53c1:01/m/44'/60'/0'/0/0"
    }],
    status: "Trezor v2.3.2 'te_da_cuen' online",
    url: "trezor://1209:53c1:01"
}]
```

We can potentially allow URIs of format similar to `"trezor://1209:53c1:01/m/44'/60'/0'/0/0"`, where the indices are variable depending on the account that is derived. In any case, I think we can do more than assuming the only valid URI is `trezor`.